cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -38,8 +42,10 @@
   private static final long serialVersionUID = -1173500180640174909L;
 
   private Map<String, Map<String, Long>> mapReduceCounters;
-  private final Map<Scope, Map<String, List<NodeValue>>> tokenValueMap = new EnumMap<>(Scope.class);
+  private Map<Scope, Map<String, List<NodeValue>>> tokenValueMap =
+    Collections.synchronizedMap(new EnumMap<Scope, Map<String, List<NodeValue>>>(Scope.class));
```

Comment:
Updated the code - 
1. While creating the entry in the enum map, now adding the instance of the `ConcurrentHashMap` as a value.
2. Creating the `Collections.synchronizedList` for the NodeValue list.
3. Created private static member in the BasicWorkflowToken, which is now used to synchronized where we iterate over the list.
4. final is in place now. I removed it because in the copy constructor, I was assigning the `tokenValueMap` to the `other.tokenValueMap` in order to use the same instance. However we can still keep the outer map different and have the same instance of the inner maps.